Source/Particles/Deposition/ChargeDeposition.H

```diff
@@ -201,6 +201,316 @@ void doChargeDepositionShapeN (const GetParticlePosition& GetPosition,
+#endif
+    const amrex::Real zmin = xyzmin[2];
+
+    amrex::Array4<amrex::Real> const& rho_arr = rho_fab.array();
+    auto rho_box = rho_fab.box();
+    amrex::IntVect const rho_type = rho_box.type();
+
+    constexpr int NODE = amrex::IndexType::NODE;
+    constexpr int CELL = amrex::IndexType::CELL;
+
+    // Loop over particles and deposit into rho_fab
+#if defined(WARPX_USE_GPUCLOCK)
+    amrex::Real* cost_real = nullptr;
+    if( load_balance_costs_update_algo == LoadBalanceCostsUpdateAlgo::GpuClock) {
+        cost_real = (amrex::Real *) amrex::The_Managed_Arena()->alloc(sizeof(amrex::Real));
+        *cost_real = 0.;
+    }
+#endif
+
+#if defined(AMREX_USE_GPU)
+    const auto dxiarr = geom.InvCellSizeArray();
+    const auto plo = geom.ProbLoArray();
+    const auto domain = geom.Domain();
+
+    const amrex::Box sample_tbox(IntVect(AMREX_D_DECL(0, 0, 0)), a_tbox_max_size - 1);
+    amrex::Box sample_tbox_x = convert(sample_tbox, ix_type);
+
+    sample_tbox_x.grow(depos_order);
+
+    const auto npts = sample_tbox_x.numPts();
+
+    const int nblocks = a_bins.numBins();
+    const auto offsets_ptr = a_bins.offsetsPtr();
+    const int threads_per_block = 256;
+
+    std::size_t shared_mem_bytes = npts*sizeof(amrex::Real);
+    amrex::IntVect bin_size = WarpX::sort_bin_size;
+
+    const std::size_t max_shared_mem_bytes = amrex::Gpu::Device::sharedMemPerBlock();
+
+    if (shared_mem_bytes > max_shared_mem_bytes) {
+      std::cout << "Tile size for charge deposition of " << shared_mem_bytes
+                << " bytes would not fit in GPU's shared memory of " << max_shared_mem_bytes << " bytes" << std::endl;
+      amrex::Abort("Tile size too big for GPU shared memory charge deposition");
+    }
+    ignore_unused(np_to_deposit);
+#endif
+
+#if defined(AMREX_USE_GPU)
+    amrex::launch(
+                  nblocks, threads_per_block, shared_mem_bytes, amrex::Gpu::gpuStream(),
+                  [=] AMREX_GPU_DEVICE () noexcept
+#else // defined(AMREX_USE_GPU)
+    amrex::ParallelFor(np_to_deposit, [=] AMREX_GPU_DEVICE (long ip) noexcept
+#endif
+      {
+#if defined(AMREX_USE_GPU)
+        const int bin_id = blockIdx.x;
+        const unsigned int bin_start = offsets_ptr[bin_id];
+        const unsigned int bin_stop = offsets_ptr[bin_id+1];
+
+        if (bin_start == bin_stop) { return; }
+
+        amrex::Box buffer_box;
+        {
+          ParticleReal xp, yp, zp;
+          GetPosition(bin_start, xp, yp, zp);
+#if (AMREX_SPACEDIM == 3)
+          IntVect iv = IntVect(int(amrex::Math::floor((xp-plo[0])*dxiarr[0])),
+                               int(amrex::Math::floor((yp-plo[1])*dxiarr[1])),
+                               int(amrex::Math::floor((zp-plo[2])*dxiarr[2])));
+#elif (AMREX_SPACEDIM == 2)
+          IntVect iv = IntVect(
+                               int(amrex::Math::floor((xp-plo[0])*dxiarr[0])),
+                               int(amrex::Math::floor((zp-plo[1])*dxiarr[1])));
+#endif
```

Comment:
There's an AMReX macro for this pattern, but it won't be applicable in WarpX after the transition to `AMREX_SPACEDIM == 3` always. Does WarpX need its own equivalent replacements?

Reply:
It's actually a slightly different pattern than the one AMReX supports - note that in the second line of the 2D case, we use `z` and dimension 1 rather than `y`. I agree we need a general solution here once we've made the switch to always 3D.